cpp/src/stream_compaction/distinct_count.cu

```diff
@@ -76,97 +60,198 @@ struct check_for_nan {
+  row_equality_comparator comp(nullate::DYNAMIC{cudf::has_nulls(input_table_view)},
+                               *table_ptr,
+                               *table_ptr,
+                               null_equality::EQUAL);
+
+  return thrust::count_if(
+    rmm::exec_policy(stream),
+    thrust::counting_iterator<cudf::size_type>(0),
+    thrust::counting_iterator<cudf::size_type>(num_rows),
+    [count_nulls, nan_is_null, should_check_nan, device_view, comp] __device__(cudf::size_type i) {
+      auto const is_null = device_view.is_null(i);
+      auto const is_nan  = nan_is_null and should_check_nan
+                             ? cudf::type_dispatcher(device_view.type(), check_nan{}, device_view, i)
+                             : false;
+      if (not count_nulls and (is_null or (nan_is_null and is_nan))) { return false; }
+      if (i == 0) { return true; }
+      if (count_nulls and nan_is_null and (is_nan or is_null)) {
+        auto const prev_is_nan =
+          should_check_nan
+            ? cudf::type_dispatcher(device_view.type(), check_nan{}, device_view, i - 1)
+            : false;
+        return not(device_view.is_null(i - 1) or prev_is_nan);
+      }
+      return not comp(i, i - 1);
+    });
+}
+
+cudf::size_type unordered_distinct_count(column_view const& input,
+                                         null_policy null_handling,
+                                         nan_policy nan_handling,
+                                         rmm::cuda_stream_view stream)
+{
+  if (0 == input.size() or input.null_count() == input.size()) { return 0; }
 
-  bool has_nan = false;
   // Check for Nans
   // Checking for nulls in input and flag nan_handling, as the count will
-  // only get affected if these two conditions are true. NAN will only be
+  // only get affected if these two conditions are true. NaN will only be
   // be an extra if nan_handling was NAN_IS_NULL and input also had null, which
   // will increase the count by 1.
-  if (input.has_nulls() and nan_handling == nan_policy::NAN_IS_NULL) {
-    has_nan = cudf::type_dispatcher(input.type(), has_nans{}, input, stream);
-  }
+  auto const has_nan_as_null = (nan_handling == nan_policy::NAN_IS_NULL) and
+                               cudf::type_dispatcher(input.type(), has_nans{}, input, stream);
+  auto const has_null = input.has_nulls();
 
-  auto count = detail::distinct_count(table_view{{input}}, null_equality::EQUAL, stream);
+  auto count = detail::unordered_distinct_count(table_view{{input}}, null_equality::EQUAL, stream);
 
   // if nan is considered null and there are already null values
-  if (nan_handling == nan_policy::NAN_IS_NULL and has_nan and input.has_nulls()) --count;
-
-  if (null_handling == null_policy::EXCLUDE and input.has_nulls())
-    return --count;
-  else
-    return count;
+  if (has_nan_as_null and (has_null or null_handling == null_policy::EXCLUDE)) { --count; }
+  if (null_handling == null_policy::EXCLUDE and has_null) { --count; }
```

Comment:
Can we put this case of `EXCLUDE and has_null` first, before we deal with the double-counted NaN-as-null? The comment `// if nan is considered null and there are already null values` doesn't apply to this line.